### PR TITLE
Move change logic to coin_select

### LINKED ISSUE #147: Move change creation/addition inside `CoinSelectionAlgorithm`
See https://github.com/bitcoindevkit/bdk/issues/121#issuecomment-716723002 for why that would be preferred.

We could simply tell the coin selection algorithm how much the change output costs in terms of weight units, and then we would get back a boolean from the coin selection that tells the wallet whether it should actually add the change or not.

### PR DESCRIPTION
**Pull request by csralvall**
### Description

The former way to compute and create change was inside `create_tx`, just after
performing coin selection.
It blocked the opportunity to have an "ensemble" algorithm to decide between
multiple coin selection algorithms based on a metric, like Waste.
Now, change is not created inside `coin_select` but the change amount and the
possibility to create change is decided inside the `coin_select` method. In
this way, change is associated with the coin selection algorithm that generated
it, and a method to decide between them can be implemented.

Fixes #147.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature
* [x] I've updated `CHANGELOG.md`

### CODE REVIEW

File: src/wallet/coin_selection.rs [hunk 1]
```diff
@@ -165,6 +186,7 @@ pub trait CoinSelectionAlgorithm<D: Database>: std::fmt::Debug {
         fee_rate: FeeRate,
         amount_needed: u64,
         fee_amount: u64,
+        _drain_output: &TxOut,
```
Comment (danielabrozzoni) [inline src/wallet/coin_selection.rs]: Since this is the trait definition I wouldn't use the `_`, calling the parameter `drain_output`. Also, drain_output needs to be documented

File: src/wallet/coin_selection.rs [hunk 2]
```diff
@@ -296,16 +321,31 @@ fn select_sorted_utxos(
...
     }
 
+    let remaining_amount = selected_amount - amount_needed_with_fees;
+
+    let change_fee = fee_rate.fee_vb(serialize(drain_output).len());
+
+    let drain_val = remaining_amount.saturating_sub(change_fee);
+
+    let excess = if drain_val.is_dust(&drain_output.script_pubkey) {
+        let target = drain_output.script_pubkey.dust_value().as_sat();
```
Comment (danielabrozzoni) [inline src/wallet/coin_selection.rs]: What is target, exactly? The target you should reach for creating a change?

File: src/wallet/coin_selection.rs [hunk 3]
```diff
@@ -111,13 +118,26 @@ pub type DefaultCoinSelectionAlgorithm = LargestFirstCoinSelection; // make the
 // prev_txid (32 bytes) + prev_vout (4 bytes) + sequence (4 bytes) + script_len (1 bytes)
 pub(crate) const TXIN_BASE_WEIGHT: usize = (32 + 4 + 4 + 1) * 4;
 
+#[derive(Debug)]
+/// Remaining amount after performing coin selection
+pub enum Excess {
+    /// It's not possible to create spendable output from excess using the current drain output
+    /// NoChange(target, remaining_amount, change_output_fee)
+    NoChange(u64, u64, u64),
```
Comment (danielabrozzoni) [inline src/wallet/coin_selection.rs]: Maybe it would be better to have here and below a struct, instead of a tuple? With 3 elements of the same type, it starts to be difficult to remember which is what
OTOH, structs are more verbose and also you need to find a name for them.

**Review by danielabrozzoni — COMMENTED** [review]
Concept ACK, it looks good to me! Just a couple of comments.

File: src/wallet/coin_selection.rs (hunk 3 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: What do you think of variants with named fields, like [this](https://users.rust-lang.org/t/enum-field-expressions/2411/6).
It can be a middle ground between both options.
E.g.:
```rust
enum Excess {
      NoChange { dust_threshold: u64, remaining_amount: u64, change_fee: u64 },
      Change { amount: u64, fee: u64 }
}
```

File: src/wallet/coin_selection.rs (hunk 2 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: The idea was that you must _hit_, at least, the dust value to create change (_the target_) . Now it doesn't seem so accurate. Maybe `dust_threshold` is.

File: src/wallet/coin_selection.rs (hunk 3 above)
Comment (danielabrozzoni) [reply src/wallet/coin_selection.rs]: That would work great!

File: src/wallet/coin_selection.rs (hunk 2 above)
Comment (danielabrozzoni) [reply src/wallet/coin_selection.rs]: Yeah `dust_threshold` is definitely easier to understand!

File: src/wallet/coin_selection.rs (hunk 3 above)
Comment (afilini) [reply src/wallet/coin_selection.rs]: ACK

(This is in reply to Daniela's comment about using named enum variants)

File: src/wallet/coin_selection.rs [hunk 4]
```diff
@@ -597,6 +638,7 @@ impl BranchAndBoundCoinSelection {
         CoinSelectionResult {
             selected,
             fee_amount,
+            excess: Excess::Change(0, 0),
```
Comment (afilini) [inline src/wallet/coin_selection.rs]: I suppose these are just temporary values, right?

File: src/wallet/coin_selection.rs (hunk 1 above)
Comment (afilini) [inline src/wallet/coin_selection.rs]: Technically here you don't need the full `TxOut`, the `Script` should be enough and I think it would be less confusing because it makes more sense that your coin selection depends on the change script, not some kind of output.

I would change this to a `drain_script` or something, then you'll have to replace the size estimation for the txout and it should be `serialize(script).len() + 8` (please double check this).

It's very important that you `serialize` the script and avoid using `Script::len()`: the former should also prefix the script with its length in VarInt format, which I believe is how it gets encoded in a `TxOut`. The latter only tells you the length of the script.

**Review by afilini — COMMENTED** [review]
Concept ACK, this looks very similar to what I had in mind in terms of architecture

File: src/wallet/coin_selection.rs (hunk 1 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: I'll change it by `drain_script: &Script` (with docs and no `_`) as @afilini suggested.

File: src/wallet/coin_selection.rs (hunk 4 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: Yes, you're right. I'm trying to find a solution with a good balance between the parameter increment and the generality to reach each of the possible outputs of `BranchAndBound`.

File: src/wallet/coin_selection.rs (hunk 1 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: Totally, this is great!
I double checked that `serialize(drain_output).len() == (serialize(drain_script).len() + 8)`.
Both are equal to: `size(len(script_pubkey)) + len(script_pubkey) + size(output_value)` where
`size(len(script_pubkey))` is the size of the `VarInt` field used to describe the length of the script.
So, it's safe for use.

**Comment by danielabrozzoni** [comment]
Can you rebase on master to pick up the CI fixes?

File: src/wallet/coin_selection.rs [hunk 5]
```diff
@@ -541,22 +628,40 @@ impl BranchAndBoundCoinSelection {
...
-            .collect();
+            .collect::<Vec<OutputGroup>>();
+
+        let selected_amount = selected_utxos
+            .iter()
+            .map(|u| u.effective_value)
+            .sum::<i64>();
+
+        // remaining_amount can't be negative as that would mean the
+        // selection wasn't successfull
```
Comment (danielabrozzoni) [inline src/wallet/coin_selection.rs]: nit: `successful` (here and in other comments as well)

File: src/wallet/coin_selection.rs [hunk 6]
```diff
@@ -541,22 +628,40 @@ impl BranchAndBoundCoinSelection {
...
         Ok(BranchAndBoundCoinSelection::calculate_cs_result(
             selected_utxos,
             required_utxos,
             fee_amount,
+            excess,
         ))
     }
 
+    #[allow(clippy::too_many_arguments)]
+    // REVIEW: Other approaches required more parameters or a large number of
```
Comment (danielabrozzoni) [inline src/wallet/coin_selection.rs]: Related to this comment: maybe we should move the change_fee calculation inside of `decide_change`? What do you think? 
```diff
diff --git a/src/wallet/coin_selection.rs b/src/wallet/coin_selection.rs
index a9ca356..720a9e4 100644
--- a/src/wallet/coin_selection.rs
+++ b/src/wallet/coin_selection.rs
@@ -300,7 +300,10 @@ impl<D: Database> CoinSelectionAlgorithm<D> for OldestFirstCoinSelection {
     }
 }
 
-fn decide_change(remaining_amount: u64, change_fee: u64, drain_script: &Script) -> Excess {
+fn decide_change(remaining_amount: u64, drain_script: &Script, fee_rate: &FeeRate) -> Excess {
+    // drain_output_len = size(len(script_pubkey)) + len(script_pubkey) + size(output_value)
+    let drain_output_len = serialize(drain_script).len() + 8usize;
+    let change_fee = fee_rate.fee_vb(drain_output_len);
     let drain_val = remaining_amount.saturating_sub(change_fee);
 
     if drain_val.is_dust(drain_script) {
@@ -360,11 +363,7 @@ fn select_sorted_utxos(
 
     let remaining_amount = selected_amount - amount_needed_with_fees;
 
-    // drain_output_len = size(len(script_pubkey)) + len(script_pubkey) + size(output_value)
-    let drain_output_len = serialize(drain_script).len() + 8usize;
-    let change_fee = fee_rate.fee_vb(drain_output_len);
-
-    let excess = decide_change(remaining_amount, change_fee, drain_script);
+    let excess = decide_change(remaining_amount, drain_script, &fee_rate);
 
     Ok(CoinSelectionResult {
         selected,
@@ -472,9 +471,6 @@ impl<D: Database> CoinSelectionAlgorithm<D> for BranchAndBoundCoinSelection {
             .try_into()
             .expect("Bitcoin amount to fit into i64");
 
-        // drain_output_len = size(len(script_pubkey)) + len(script_pubkey) + size(output_value)
-        let drain_output_len = serialize(drain_script).len() + 8usize;
-        let change_fee = fee_rate.fee_vb(drain_output_len);
 
         if curr_value > actual_target {
             let selected_amount = required_utxos
@@ -486,7 +482,7 @@ impl<D: Database> CoinSelectionAlgorithm<D> for BranchAndBoundCoinSelection {
             // actual_target = amount_needed + (fee_amount - vin_fees)
             let remaining_amount = (selected_amount - actual_target) as u64;
 
-            let excess = decide_change(remaining_amount, change_fee, drain_script);
+            let excess = decide_change(remaining_amount, drain_script, &fee_rate);
 
             return Ok(BranchAndBoundCoinSelection::calculate_cs_result(
                 vec![],
@@ -506,7 +502,7 @@ impl<D: Database> CoinSelectionAlgorithm<D> for BranchAndBoundCoinSelection {
                 fee_amount,
                 cost_of_change,
                 drain_script,
-                change_fee,
+                &fee_rate,
             )
             .unwrap_or_else(|_| {
                 self.single_random_draw(
@@ -516,7 +512,7 @@ impl<D: Database> CoinSelectionAlgorithm<D> for BranchAndBoundCoinSelection {
                     actual_target,
                     fee_amount,
                     drain_script,
-                    change_fee,
+                    &fee_rate,
                 )
             }))
     }
@@ -536,7 +532,7 @@ impl BranchAndBoundCoinSelection {
         fee_amount: u64,
         cost_of_change: f32,
         drain_script: &Script,
-        change_fee: u64,
+        fee_rate: &FeeRate,
     ) -> Result<CoinSelectionResult, Error> {
         // current_selection[i] will contain true if we are using optional_utxos[i],
         // false otherwise. Note that current_selection.len() could be less than
@@ -640,7 +636,7 @@ impl BranchAndBoundCoinSelection {
         // actual_target = amount_needed + (fee_amount - vin_fees)
         let remaining_amount = (selected_amount - actual_target) as u64;
 
-        let excess = decide_change(remaining_amount, change_fee, drain_script);
+        let excess = decide_change(remaining_amount, drain_script, fee_rate);
 
         Ok(BranchAndBoundCoinSelection::calculate_cs_result(
             selected_utxos,
@@ -661,7 +657,7 @@ impl BranchAndBoundCoinSelection {
         actual_target: i64,
         fee_amount: u64,
         drain_script: &Script,
-        change_fee: u64,
+        fee_rate: &FeeRate,
     ) -> CoinSelectionResult {
         #[cfg(not(test))]
         optional_utxos.shuffle(&mut thread_rng());
@@ -694,7 +690,7 @@ impl BranchAndBoundCoinSelection {
         // actual_target = amount_needed + (fee_amount - vin_fees)
         let remaining_amount = (selected_amount - actual_target) as u64;
 
-        let excess = decide_change(remaining_amount, change_fee, drain_script);
+        let excess = decide_change(remaining_amount, drain_script, fee_rate);
 
         BranchAndBoundCoinSelection::calculate_cs_result(
             selected_utxos,
```

This means that decide_change, bnb and single_random_draw will need a fee_rate parameter - but also that we don't have to repeat the `drain_output_len` and `change_fee` calculation in various parts of the code (this could come really handy when we separate SRD and BNB!).

**Review by danielabrozzoni — COMMENTED** [review]
This is looking really good! I have a small comment about `decide_change`, otherwise it's good to go

Edit: also it needs a rebase for conflicts

File: src/wallet/coin_selection.rs (hunk 6 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: I agree. Just a little gotcha, considering an x86_64 architecture, `FeeRate` is a `f32` so it takes 4 bytes and implements the copy trait. A reference size is `usize`, that is 8 bytes. Should we prefer references over moves or the other way around?

File: src/wallet/coin_selection.rs (hunk 5 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: Fixed.

File: src/wallet/coin_selection.rs (hunk 6 above)
Comment (danielabrozzoni) [reply src/wallet/coin_selection.rs]: You're right, FeeRate also implements Copy - let's avoid the reference and just copy it :)

Comment (csralvall) [reply src/wallet/coin_selection.rs]: Done.

**Review by danielabrozzoni — COMMENTED** [review]
Can you squash the two commits together plz? Then I'll ACK :)

File: CHANGELOG.md [hunk 7]
```diff
@@ -27,6 +27,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - Support for `tr()` descriptors in the `descriptor!()` macro
 - Add support for Bitcoin Core 23.0 when using the `rpc` blockchain
 - Add `remove_partial_sigs` and `try_finalize` to `SignOptions`
+- Add `Excess` enum to handle remaining amount after coin selection.
```
Comment (danielabrozzoni) [inline CHANGELOG.md]: These two should probably go under "Unreleased"

Comment (csralvall) [reply CHANGELOG.md]: Corrected.

**Review by danielabrozzoni — APPROVED** [review]
tACK b0f6a2089b44c3b18e97f8c2bf0183f65c67e958

File: src/wallet/coin_selection.rs [hunk 8]
```diff
@@ -42,6 +43,7 @@
 //!         fee_rate: FeeRate,
 //!         amount_needed: u64,
 //!         fee_amount: u64,
+//!         _drain_script: &Script,
```
Comment (vladimirfomene) [inline src/wallet/coin_selection.rs]: Forgot to remove the `_`, I think.

File: src/wallet/coin_selection.rs [hunk 9]
```diff
@@ -261,7 +296,28 @@ impl<D: Database> CoinSelectionAlgorithm<D> for OldestFirstCoinSelection {
                 .chain(optional_utxos.into_iter().map(|utxo| (false, utxo)))
         };
 
-        select_sorted_utxos(utxos, fee_rate, amount_needed, fee_amount)
+        select_sorted_utxos(utxos, fee_rate, amount_needed, fee_amount, drain_script)
+    }
+}
+
+fn decide_change(remaining_amount: u64, fee_rate: FeeRate, drain_script: &Script) -> Excess {
+    // drain_output_len = size(len(script_pubkey)) + len(script_pubkey) + size(output_value)
```
Comment (vladimirfomene) [inline src/wallet/coin_selection.rs:313]: From the comment for the `drain_output_len` computation, I was expecting, a third component in the form of the size of the `script_pubkey` in the computation below. I only see the addition of two components. Is there something, I'm missing?

**Review by vladimirfomene — COMMENTED** [review]
tACK. Just Nits

File: src/wallet/coin_selection.rs (hunk 9 above)
Comment (afilini) [reply src/wallet/coin_selection.rs:313]: When you `serialize` a `Script` rust-bitcoin prefixes it with the length (which is a VarInt), so the first term amounts for both `serialize(len(script_pubkey)) + len(script_pubkey)`.

As I explained in the other comment, using `Script::len()` would have only returned the length of the script itself, without accounting for the VarInt preceeding it.

But the way this is done should be correct.

Also check out: https://github.com/bitcoindevkit/bdk/pull/630#discussion_r906696742

File: src/wallet/coin_selection.rs (hunk 8 above)
Comment (danielabrozzoni) [reply src/wallet/coin_selection.rs]: It's fine to use it here, as we're not using the drain_script anywhere in this function :)

File: src/wallet/coin_selection.rs [hunk 10]
```diff
@@ -69,6 +71,7 @@
 //!         Ok(CoinSelectionResult {
 //!             selected: all_utxos_selected,
 //!             fee_amount: fee_amount + additional_fees,
+//!             excess: Excess::Change { amount: 0, fee: 0 },
```
Comment (afilini) [inline src/wallet/coin_selection.rs]: This isn't the best in an example, because it probably creates an invalid tx (zero-value output for the change)

File: src/wallet/coin_selection.rs [hunk 11]
```diff
@@ -408,10 +472,22 @@ impl<D: Database> CoinSelectionAlgorithm<D> for BranchAndBoundCoinSelection {
             .expect("Bitcoin amount to fit into i64");
 
         if curr_value > actual_target {
+            let selected_amount = required_utxos
+                .iter()
+                .map(|u| u.effective_value)
+                .sum::<i64>();
```
Comment (afilini) [inline src/wallet/coin_selection.rs]: Isn't this literally `curr_value`? Why do you have to recompute it here?

File: src/wallet/coin_selection.rs [hunk 12]
```diff
@@ -579,13 +677,31 @@ impl BranchAndBoundCoinSelection {
             })
             .collect::<Vec<_>>();
 
-        BranchAndBoundCoinSelection::calculate_cs_result(selected_utxos, required_utxos, fee_amount)
+        let selected_amount = selected_utxos
+            .iter()
+            .map(|u| u.effective_value)
+            .sum::<i64>();
```
Comment (afilini) [inline src/wallet/coin_selection.rs]: You can probably merge this into the previous lines, replacing the `scan` with a `fold`. Keep a `(u64, Vec<_>)` as your internal state, every time you pick one utxo add the value to the accumulator and push it to the vec.

At the end you have both the selected utxos and the sum of their values

**Review by afilini — COMMENTED** [review]
A few small comments.

Also I was thinking: do you think we should add some tests for the coin selection? I noticed that most of the tests were just updated to assume an empty drain_script and I'm actually a bit suprised none of the tests started failing (I was assuming at least some tests were asserting conditions on BDK not adding a change output, and reducing the size of the drain_script should cause them to potentially start adding one).

Can you double check if we are missing something there maybe?

File: src/wallet/coin_selection.rs (hunk 10 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: Yes, I missed that. I can make `decide_change` public to compute the excess, but that would imply that it needs to be documented (I don't know if the function it's useful enough to be exposed in the docs) or I can mimic its behavior inside `AlwaysSpendEverything`.

File: src/wallet/coin_selection.rs (hunk 11 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: So true!

File: src/wallet/coin_selection.rs (hunk 12 above)
Comment (csralvall) [reply src/wallet/coin_selection.rs]: Working on it, thanks!

**Comment by csralvall** [comment]
@afilini 
> Also I was thinking: do you think we should add some tests for the coin selection? I noticed that most of the tests were just updated to assume an empty drain_script and I'm actually a bit suprised none of the tests started failing (I was assuming at least some tests were asserting conditions on BDK not adding a change output, and reducing the size of the drain_script should cause them to potentially start adding one).

They are not failing because the former code didn't make any change computations inside `coin_selection`, so the `excess` field who stores the change isn't touched by any of the tests inside `coin_selection.rs`. However, the correctness of change results is tested in the tests inside "mod", as the `decide_change` output is directly used there. Not the best option in terms of architecture, but at the beginning I thought they were enough.
That said, while reviewing your comments, I discovered the `selected_amount` I was using to compute the `remaining_amount` and then the change, was not considering the `required_utxos` amount in the BranchAndBoundAlgorithm and the tests didn't throw any error.
So, I'm in favor of adding or improving the tests. Like @rajarshimaitra said [here](https://github.com/bitcoindevkit/bdk/pull/666#pullrequestreview-1047902278), we may not be adequately testing the fee situations and edge cases.

> Can you double check if we are missing something there maybe?

So yes, I'm going to edit those empty scripts and see how I can improve the tests.

**Comment by csralvall** [comment]
I've pushed some changes in this [branch](https://github.com/csralvall/bdk/commit/94951d4caa3e8b7068151a5d08a46a4b1af7c01a) underpinning the view that my proposal for the `coin_selection` module would have in the future.
The code is to display my ideas, so it won't compile yet.
I think that the return value of `coin_select` in that branch might be the final version. So, I'm not sure if it's worthy, at the end, to include the tests that I mentioned, since the functionality will be moved outside `coin_select` to a more general function.
I made some attempts to include tests, but all my ideas fell in circular logic, resembling highly to the `decide_change` function, already inside tested code.
The current checks and the fact that it's a refactoring and not an enhancement, are telling us that the PR didn't broke any current functionality. If something is wrong (and isn't reported by checks), it must be due to something previous, so a new PR addressing exclusively those errors and edge cases, or the tests for wallet in general, may be more appropriate.
With all this I mean that I've reviewed the current state of tests and any new change to the the modified code won't improve them any further.

**Review by afilini — APPROVED** [review]
ACK 32ae95f463f62c42c6d6aec62c1832a30298fce4
